### DISCUSSION TITLE
Stress calculation support

### DISCUSSION THREAD

**Discussion by youngjae142** [mir-group/pair_nequip#14] [Ideas]
Hello!
I'm trying to calculate stress and virial with lammps.
I noticed that there is a stress-develop branch in the nequip repository.
It would be awesome if it is implemented in pair_nequip and nequip-deploy too! 😀

**Comment by Linux-cpp-lisp** [comment]
Hi @ssrokyz ,

:smile: thanks for your ongoing interest in our code!

We do have the `stress-develop` branch, which is almost ready to be merged pending some final testing. If you would like to be among the "beta testers" it is definitely ready to start try using— just please sanity check and verify your results and let us know if you encounter anything you think may be a bug with the code.

We also plan to implement stress support in `pair_nequip` at some point soon, so stay tuned.

**Reply by youngjae142** [reply]
Thx, @Linux-cpp-lisp !
I just successfully installed the stress branch. I'll test it and let you know.
By the way, if you need someone to test the pair_nequip someday, I'm all ready!
I'm aware of viral and stress theories and used to calculate the Green-Kubo relation (lattice thermal conductivity) with lammps.
Have a nice day!

**Reply by Linux-cpp-lisp** [reply]
@ssrokyz just wanted to point you to today's discussion in https://github.com/mir-group/nequip/issues/114 on this subject.

**Comment by youngjae142** [comment]
@Linux-cpp-lisp 
Thank you for letting me know!
I'm following all activities of NequIP and Pair NequIP these days while we are using NequIP in our Lab.
Your work on stress and virial implementation to the ASE calculator is cool!

It would be great if "atomic" virial calculation is also implemented.
It is necessary to calculate `atomic virial (W_i)` to calculate the heat flux kinds of things [heat/flux reference below].
(I was wondering if pair nequip is capable of heat flux calculation, so I tested it, but the outputs were incorrect of course. It was a meaningless fluctuation, and I think it was a free-particle stress contribution, which is `m*v_a*v_b` term.)
FYI, `(atomic stress S_i) = (atomic virial W_i) + (free-particle contribution m*v_a*v_b)` [stress/atom reference]

I'm no expert on CPP language and LAMMPS codes, so I'm not sure, I think the necessary job is to implement atomic virial in pair_nequip.cpp file which is `W_ab` term.
I've previously tested heat flux calculation with the Deep Potential package of version 1.3.3, which has definitions of atomic virial in pair_nnp.cpp file, and it worked good, FYI.

I'm always grateful for your work. Have a nice weekend. 😊

References
[stress/atom calculation of LAMMPS](https://docs.lammps.org/compute_stress_atom.html)
[heat/flux calculation of LAMMPS](https://docs.lammps.org/compute_heat_flux.html)
